### PR TITLE
cpu/sam0_common: avoid bitfield usage

### PR DESCRIPTION
**Pull request by dylad**
### Contribution description

This PR removes all bitfields usage from `cpu/sam0_common/` folder.
I decided to go with a commit per file to allow an easy revert in case something got terribly wrong. Nevertheless, CI should catch most of the issues early. However, I might get sleepy here and there so please, take a careful eyes at this.

/!\ **Breakages are highly probable here** /!\

### Testing procedure
Triple careful review and CI should save us hopefully.

### Issues/PRs references
#20457

### CODE REVIEW

File: cpu/sam0_common/periph/adc.c [hunk 1]
```diff
@@ -190,10 +190,10 @@ static int _adc_configure(Adc *dev, adc_res_t res)
     /* Set ADC resolution */
 #ifdef ADC_CTRLC_RESSEL
     /* Reset resolution bits in CTRLC */
-    dev->CTRLC.bit.RESSEL = res & 0x3;
+    dev->CTRLC.reg |= ADC_CTRLC_RESSEL(res);
 #else
     /* Reset resolution bits in CTRLB */
-    dev->CTRLB.bit.RESSEL = res & 0x3;
+    dev->CTRLB.reg |= ADC_CTRLB_RESSEL(res);
```
Comment (dylad) [inline cpu/sam0_common/periph/adc.c:196]: This needs a `foo &= ~bar`

File: cpu/sam0_common/periph/dma.c [hunk 2]
```diff
@@ -255,11 +255,11 @@ void dma_start(dma_t dma)
 
 #ifdef REG_DMAC_CHID
     unsigned state = irq_disable();
-    DMAC->CHID.bit.ID = dma;
+    DMAC->CHID.reg |= DMAC_CHID_ID(dma);
```
Comment (dylad) [inline cpu/sam0_common/periph/dma.c:258]: This needs a `foo &= ~bar`

File: cpu/sam0_common/periph/dma.c [hunk 3]
```diff
@@ -274,15 +274,15 @@ void dma_cancel(dma_t dma)
     DEBUG("[DMA]: Cancelling active transfer: %u\n", dma);
 #ifdef REG_DMAC_CHID
     unsigned state = irq_disable();
-    DMAC->CHID.bit.ID = dma;
+    DMAC->CHID.reg |= DMAC_CHID_ID(dma);
```
Comment (dylad) [inline cpu/sam0_common/periph/dma.c:277]: This needs a `foo &= ~bar`

File: cpu/sam0_common/include/periph_cpu_common.h [hunk 4]
```diff
@@ -681,8 +681,9 @@ typedef enum {
 static inline void sam0_set_voltage_regulator(sam0_supc_t src)
 {
 #ifdef REG_SUPC_VREG
-    SUPC->VREG.bit.SEL = src;
-    while (!SUPC->STATUS.bit.VREGRDY) {}
+    SUPC->VREG.reg &= ~(1 << SUPC_VREG_SEL_Pos);
+    SUPC->VREG.reg |= (src << SUPC_VREG_SEL_Pos);
```
[the remark this answers is not in the document]
Comment (kfessel) [reply cpu/sam0_common/include/periph_cpu_common.h:685]: T = x is one write 
T |= is a read modify write 

write without read is better than one rmw 

the old one sadly was a rmw  due to being a bitfield access ( unless there are special magic instructions like on the avr (out)))

File: cpu/sam0_common/periph/usbdev.c [hunk 5]
```diff
@@ -210,19 +212,25 @@ static bool _ep_in_flags_set(UsbDeviceEndpoint *ep_reg)
 
 static void _set_address(sam0_common_usb_t *dev, uint8_t addr)
 {
-    dev->config->device->DADD.bit.DADD = addr;
+    dev->config->device->DADD.reg &= ~USB_DEVICE_DADD_DADD_Msk;
+    dev->config->device->DADD.reg |= USB_DEVICE_DADD_DADD(addr);
```
Comment (kfessel) [inline cpu/sam0_common/periph/usbdev.c:216]: making 1 rmw into 2 rmw?
temporary

File: cpu/sam0_common/periph/usbdev.c [hunk 6]
```diff
@@ -469,10 +479,12 @@ static void _ep_enable(usbdev_ep_t *ep)
...
     }
     if (ep->dir == USB_EP_DIR_OUT) {
-        ep_reg->EPCFG.bit.EPTYPE0 = type;
+        ep_reg->EPCFG.reg &= ~USB_DEVICE_EPCFG_EPTYPE0_Msk;
+        ep_reg->EPCFG.reg |= USB_DEVICE_EPCFG_EPTYPE0(type);
     }
     else {
-        ep_reg->EPCFG.bit.EPTYPE1 = type;
+        ep_reg->EPCFG.reg &= ~USB_DEVICE_EPCFG_EPTYPE1_Msk;
+        ep_reg->EPCFG.reg |= USB_DEVICE_EPCFG_EPTYPE1(type);
```
Comment (kfessel) [inline cpu/sam0_common/periph/usbdev.c:487]: making 1 rmw into 2 rmw?
temporary

File: cpu/sam0_common/periph/usbdev.c [hunk 7]
```diff
@@ -710,7 +723,8 @@ static int _usbdev_ep_xmit(usbdev_ep_t *ep, uint8_t *buf, size_t len)
     _bank_from_ep(ep)->ADDR.reg = (uint32_t)(intptr_t)buf;
     if (ep->dir == USB_EP_DIR_IN) {
         _disable_ep_stall_in(ep_reg);
-        _bank_from_ep(ep)->PCKSIZE.bit.BYTE_COUNT = len;
+        _bank_from_ep(ep)->PCKSIZE.reg &= ~USB_DEVICE_PCKSIZE_BYTE_COUNT_Msk;
+        _bank_from_ep(ep)->PCKSIZE.reg |= USB_DEVICE_PCKSIZE_BYTE_COUNT(len);
```
Comment (kfessel) [inline cpu/sam0_common/periph/usbdev.c:727]: making 1 rmw into 2 rmw?
temporary

File: cpu/sam0_common/periph/wdt.c [hunk 8]
```diff
@@ -203,7 +212,8 @@ void wdt_setup_reboot_with_callback(uint32_t min_time, uint32_t max_time,
         }
 
         WDT->INTENSET.reg = WDT_INTENSET_EW;
-        WDT->EWCTRL.bit.EWOFFSET = per - warning_offset;
+        WDT->EWCTRL.reg &= ~WDT_EWCTRL_EWOFFSET_Msk;
+        WDT->EWCTRL.reg |= WDT_EWCTRL_EWOFFSET(per - warning_offset);
```
Comment (kfessel) [inline cpu/sam0_common/periph/wdt.c:216]: making 1 rmw into 2 rmw?
temporary

**Review by kfessel — COMMENTED** [review]
there are a lot of 1 rmw to  2 rmw cases.

some might need a temporary,  can probably be some thing  like 

```
REG = (REG & ~Msk) | ((val << Pos) &Msk);
```
there might be a macro for this

**Review by kfessel — DISMISSED** [review]
Please do not merge until the doubling of the rmws is fixed

**Comment by dylad** [comment]
@kfessel I've pushed my fixup commits.

File: cpu/sam0_common/periph/timer.c [hunk 9]
```diff
@@ -332,20 +336,24 @@ int timer_clear(tim_t tim, int channel)
...
         cycles for the timer peripheral to set the SYNCBUSY/READSYNC bit
         after writing the READSYNC bit
 
         The problem was observed on SAME54.
       */
-    while(dev(tim)->CTRLBSET.bit.CMD == TC_CTRLBSET_CMD_READSYNC_Val) {}
+    do {
+        cmd = ((dev(tim)->CTRLBSET.reg & TC_CTRLBSET_CMD_Msk) >> TC_CTRLBSET_CMD_Pos);
+    } while(cmd == TC_CTRLBSET_CMD_READSYNC_Val);
+
```
Comment (kfessel) [inline cpu/sam0_common/periph/timer.c:356]: might be faster but please ignore unless someone objdumped

```suggestion
    do {
        cmd = ((dev(tim)->CTRLBSET.reg & TC_CTRLBSET_CMD_Msk));
    } while(cmd == TC_CTRLBSET_CMD_READSYNC_Val << TC_CTRLBSET_CMD_Pos);

```

File: cpu/sam0_common/periph/adc.c [hunk 10]
```diff
@@ -172,6 +172,7 @@ static void _setup_calibration(Adc *dev)
 
 static int _adc_configure(Adc *dev, adc_res_t res)
 {
+    uint32_t ctrl;
```
Comment (kfessel) [inline cpu/sam0_common/periph/adc.c:175]:
```suggestion
```

File: cpu/sam0_common/periph/adc.c [hunk 11]
```diff
@@ -190,10 +191,12 @@ static int _adc_configure(Adc *dev, adc_res_t res)
     /* Set ADC resolution */
 #ifdef ADC_CTRLC_RESSEL
     /* Reset resolution bits in CTRLC */
-    dev->CTRLC.bit.RESSEL = res & 0x3;
+    ctrl = dev->CTRLC.reg;
```
Comment (kfessel) [inline cpu/sam0_common/periph/adc.c:194]:
```suggestion
    uint32_t ctrlc = dev->CTRLC.reg;
```

File: cpu/sam0_common/periph/adc.c [hunk 12]
```diff
@@ -190,10 +191,12 @@ static int _adc_configure(Adc *dev, adc_res_t res)
...
+    dev->CTRLC.reg = ((ctrl & ~ADC_CTRLC_RESSEL_Msk) | ADC_CTRLC_RESSEL(res));
 #else
     /* Reset resolution bits in CTRLB */
-    dev->CTRLB.bit.RESSEL = res & 0x3;
+    ctrl = dev->CTRLB.reg;
```
Comment (kfessel) [inline cpu/sam0_common/periph/adc.c:198]:
```suggestion
    uint32_t ctrlb = dev->CTRLB.reg;
```

**Review by kfessel — COMMENTED** [review]
:+1: 

mostly good

File: cpu/sam0_common/periph/dma.c [hunk 13]
```diff
@@ -255,11 +255,12 @@ void dma_start(dma_t dma)
 
 #ifdef REG_DMAC_CHID
     unsigned state = irq_disable();
-    DMAC->CHID.bit.ID = dma;
+    uint32_t reg = DMAC->CHID.reg;
+    DMAC->CHID.reg = ((reg & ~DMAC_CHID_ID_Msk) | DMAC_CHID_ID(dma));
```
[the remark this answers is not in the document]
Comment (kfessel) [reply cpu/sam0_common/periph/dma.c:259]: for samd21
```suggestion
    DMAC->CHID.reg = DMAC_CHID_ID(dma));
```

File: cpu/sam0_common/periph/usbdev.c [hunk 14]
```diff
@@ -450,6 +461,7 @@ static void _ep_enable(usbdev_ep_t *ep)
           ep->dir == USB_EP_DIR_OUT ? "OUT" : "IN");
     UsbDeviceEndpoint *ep_reg = _ep_reg_from_ep(ep);
     uint8_t type = 0;
+    uint32_t reg;
```
[the remark this answers is not in the document]
Comment (dylad) [reply cpu/sam0_common/periph/usbdev.c:464]: I don't mind renaming and define it right when needed.
It's just an old habit of mine to define variable at the beginning of the function.

**Comment by dylad** [comment]
@kfessel A new round of fixups has been pushed.

File: cpu/sam0_common/periph/adc.c [hunk 15]
```diff
@@ -190,10 +190,12 @@ static int _adc_configure(Adc *dev, adc_res_t res)
...
 #ifdef ADC_CTRLC_RESSEL
     /* Reset resolution bits in CTRLC */
-    dev->CTRLC.bit.RESSEL = res & 0x3;
+    uint32_t ctrlc = dev->CTRLC.reg;
+    dev->CTRLC.reg = ((ctrlc & ~ADC_CTRLC_RESSEL_Msk) | ADC_CTRLC_RESSEL(res));
 #else
     /* Reset resolution bits in CTRLB */
-    dev->CTRLB.bit.RESSEL = res & 0x3;
+    uint32_t ctrlb = dev->CTRLB.reg;
+    dev->CTRLB.reg = ((ctrlb & ~ADC_CTRLB_RESSEL_Msk) | ADC_CTRLB_RESSEL(res));
```
Comment (kfessel) [inline cpu/sam0_common/periph/adc.c:198]: the function is called configure but does not configure all the bits (and nothing else in this file does) this seem to me like bad code translated well.

*you might ignore this comment your translation mimics the old behavior*

**Review by kfessel — COMMENTED** [review]
it getting fewer

btw i'd like to ask: if you are doing tests?

File: cpu/sam0_common/periph/adc.c [hunk 16]
```diff
@@ -311,13 +313,15 @@ static Adc *_adc(uint8_t dev)
 static int32_t _sample(adc_t line)
 {
     Adc *dev = _dev(line);
-    bool diffmode = adc_channels[line].inputctrl & ADC_INPUTCTRL_DIFFMODE;
+    bool diffmode = !!(adc_channels[line].inputctrl & ADC_INPUTCTRL_DIFFMODE);
```
Comment (benpicco) [inline cpu/sam0_common/periph/adc.c:316]:
```suggestion
    bool diffmode = adc_channels[line].inputctrl & ADC_INPUTCTRL_DIFFMODE;
```

same here

File: cpu/sam0_common/periph/uart.c [hunk 17]
```diff
@@ -461,28 +461,28 @@ bool uart_collision_detected(uart_t uart)
      */
     _syncbusy(dev(uart));
 
-    bool collision = dev(uart)->STATUS.bit.COLL;
+    bool collision = !!(dev(uart)->STATUS.reg & SERCOM_USART_STATUS_COLL);
```
Comment (benpicco) [inline cpu/sam0_common/periph/uart.c:464]:
```suggestion
    bool collision = dev(uart)->STATUS.reg & SERCOM_USART_STATUS_COLL;
```

File: cpu/sam0_common/periph/usbdev.c [hunk 18]
```diff
@@ -210,19 +212,27 @@ static bool _ep_in_flags_set(UsbDeviceEndpoint *ep_reg)
...
 static bool _syncbusy_enable(sam0_common_usb_t *dev)
 {
-    return dev->config->device->SYNCBUSY.bit.ENABLE;
+    return !!(dev->config->device->SYNCBUSY.reg & USB_SYNCBUSY_ENABLE);
 }
 
 static bool _syncbusy_swrst(sam0_common_usb_t *dev)
 {
-    return dev->config->device->SYNCBUSY.bit.SWRST;
+    return !!(dev->config->device->SYNCBUSY.reg & USB_SYNCBUSY_SWRST);
```
Comment (benpicco) [inline cpu/sam0_common/periph/usbdev.c:235]:
```suggestion
    return dev->config->device->SYNCBUSY.reg & USB_SYNCBUSY_SWRST;
```

File: cpu/sam0_common/periph/usbdev.c [hunk 19]
```diff
@@ -210,19 +212,27 @@ static bool _ep_in_flags_set(UsbDeviceEndpoint *ep_reg)
...
+    }
+
+    dev->config->device->DADD.reg = ((dadd & ~USB_DEVICE_DADD_DADD_Msk) |
+                                     USB_DEVICE_DADD_DADD(addr));
 }
 
 static bool _syncbusy_enable(sam0_common_usb_t *dev)
 {
-    return dev->config->device->SYNCBUSY.bit.ENABLE;
+    return !!(dev->config->device->SYNCBUSY.reg & USB_SYNCBUSY_ENABLE);
```
Comment (benpicco) [inline cpu/sam0_common/periph/usbdev.c:230]:
```suggestion
    return dev->config->device->SYNCBUSY.reg & USB_SYNCBUSY_ENABLE;
```

File: cpu/sam0_common/periph/usbdev.c [hunk 20]
```diff
@@ -210,19 +212,27 @@ static bool _ep_in_flags_set(UsbDeviceEndpoint *ep_reg)
...
-    dev->config->device->DADD.bit.ADDEN = addr ? 1 : 0;
+    if (addr) {
+        dadd |= USB_DEVICE_DADD_ADDEN;
+    }
+    else {
+        dadd &= ~USB_DEVICE_DADD_ADDEN;
+    }
+
+    dev->config->device->DADD.reg = ((dadd & ~USB_DEVICE_DADD_DADD_Msk) |
+                                     USB_DEVICE_DADD_DADD(addr));
```
Comment (benpicco) [inline cpu/sam0_common/periph/usbdev.c:225]:
```suggestion
    dev->config->device->DADD.reg = (dadd & ~USB_DEVICE_DADD_DADD_Msk)
                                  |  USB_DEVICE_DADD_DADD(addr);
```

File: cpu/sam0_common/periph/usbdev.c [hunk 21]
```diff
@@ -342,7 +352,10 @@ static void _usbdev_init(usbdev_t *dev)
         USB_PADCAL_TRIM((*(uint32_t *)USB_FUSES_TRIM_ADDR >>
                          USB_FUSES_TRIM_Pos));
 
-    usbdev->config->device->CTRLB.bit.SPDCONF = USB_DEVICE_CTRLB_SPDCONF_FS;
+    uint32_t ctrlb = usbdev->config->device->CTRLB.reg;
+    usbdev->config->device->CTRLB.reg = ((ctrlb & ~USB_DEVICE_CTRLB_SPDCONF_Msk) |
+                                         USB_DEVICE_CTRLB_SPDCONF_FS);
```
Comment (benpicco) [inline cpu/sam0_common/periph/usbdev.c:357]:
```suggestion
    usbdev->config->device->CTRLB.reg = (ctrlb & ~USB_DEVICE_CTRLB_SPDCONF_Msk)
                                      | USB_DEVICE_CTRLB_SPDCONF_FS;
```

File: cpu/sam0_common/periph/usbdev.c [hunk 22]
```diff
@@ -469,10 +483,14 @@ static void _ep_enable(usbdev_ep_t *ep)
         assert(false);
     }
     if (ep->dir == USB_EP_DIR_OUT) {
-        ep_reg->EPCFG.bit.EPTYPE0 = type;
+        epcfg = ep_reg->EPCFG.reg;
+        ep_reg->EPCFG.reg = ((epcfg & ~USB_DEVICE_EPCFG_EPTYPE0_Msk) |
+                             USB_DEVICE_EPCFG_EPTYPE0(type));
```
Comment (benpicco) [inline cpu/sam0_common/periph/usbdev.c:488]:
```suggestion
        ep_reg->EPCFG.reg = (epcfg & ~USB_DEVICE_EPCFG_EPTYPE0_Msk)
                           | USB_DEVICE_EPCFG_EPTYPE0(type);
```

File: cpu/sam0_common/periph/usbdev.c [hunk 23]
```diff
@@ -469,10 +483,14 @@ static void _ep_enable(usbdev_ep_t *ep)
...
     }
     else {
-        ep_reg->EPCFG.bit.EPTYPE1 = type;
+        epcfg = ep_reg->EPCFG.reg;
+        ep_reg->EPCFG.reg = ((epcfg & ~USB_DEVICE_EPCFG_EPTYPE1_Msk) |
+                             USB_DEVICE_EPCFG_EPTYPE1(type));
```
Comment (benpicco) [inline cpu/sam0_common/periph/usbdev.c:493]:
```suggestion
        ep_reg->EPCFG.reg = (epcfg & ~USB_DEVICE_EPCFG_EPTYPE1_Msk)
                           | USB_DEVICE_EPCFG_EPTYPE1(type);
```

File: cpu/sam0_common/periph/usbdev.c [hunk 24]
```diff
@@ -644,7 +662,8 @@ static void _usbdev_ep_init(usbdev_ep_t *ep)
 
 static size_t _ep_get_available(usbdev_ep_t *ep)
 {
-    return _bank_from_ep(ep)->PCKSIZE.bit.BYTE_COUNT;
+    return ((_bank_from_ep(ep)->PCKSIZE.reg & USB_DEVICE_PCKSIZE_BYTE_COUNT_Msk) >>
+            USB_DEVICE_PCKSIZE_BYTE_COUNT_Pos);
```
Comment (benpicco) [inline cpu/sam0_common/periph/usbdev.c:666]:
```suggestion
    return ((_bank_from_ep(ep)->PCKSIZE.reg & USB_DEVICE_PCKSIZE_BYTE_COUNT_Msk)
           >> USB_DEVICE_PCKSIZE_BYTE_COUNT_Pos);
```

easier to read

File: cpu/sam0_common/periph/usbdev.c [hunk 25]
```diff
@@ -710,7 +729,9 @@ static int _usbdev_ep_xmit(usbdev_ep_t *ep, uint8_t *buf, size_t len)
     _bank_from_ep(ep)->ADDR.reg = (uint32_t)(intptr_t)buf;
     if (ep->dir == USB_EP_DIR_IN) {
         _disable_ep_stall_in(ep_reg);
-        _bank_from_ep(ep)->PCKSIZE.bit.BYTE_COUNT = len;
+        uint32_t reg = _bank_from_ep(ep)->PCKSIZE.reg;
+        _bank_from_ep(ep)->PCKSIZE.reg = ((reg & ~USB_DEVICE_PCKSIZE_BYTE_COUNT_Msk) |
+                                           USB_DEVICE_PCKSIZE_BYTE_COUNT(len));
```
Comment (benpicco) [inline cpu/sam0_common/periph/usbdev.c:734]:
```suggestion
        _bank_from_ep(ep)->PCKSIZE.reg = (reg & ~USB_DEVICE_PCKSIZE_BYTE_COUNT_Msk)
                                       |  USB_DEVICE_PCKSIZE_BYTE_COUNT(len);
```

File: cpu/sam0_common/periph/wdt.c [hunk 26]
```diff
@@ -203,7 +210,9 @@ void wdt_setup_reboot_with_callback(uint32_t min_time, uint32_t max_time,
         }
 
         WDT->INTENSET.reg = WDT_INTENSET_EW;
-        WDT->EWCTRL.bit.EWOFFSET = per - warning_offset;
+        uint32_t reg = WDT->EWCTRL.reg;
+        WDT->EWCTRL.reg = ((reg & ~WDT_EWCTRL_EWOFFSET_Msk) |
+                           WDT_EWCTRL_EWOFFSET(per - warning_offset));
```
Comment (benpicco) [inline cpu/sam0_common/periph/wdt.c:215]:
```suggestion
        WDT->EWCTRL.reg = (reg & ~WDT_EWCTRL_EWOFFSET_Msk)
                        | WDT_EWCTRL_EWOFFSET(per - warning_offset);
```

File: cpu/sam0_common/sam0_sdhc/sdhc.c [hunk 27]
```diff
@@ -106,14 +106,19 @@ static bool sdio_test_type(sdhc_state_t *state);
 
 static bool _card_detect(sdhc_state_t *state)
 {
-    return state->dev->PSR.bit.CARDINS;
+    return !!(state->dev->PSR.reg & SDHC_PSR_CARDINS);
```
Comment (benpicco) [inline cpu/sam0_common/sam0_sdhc/sdhc.c:109]:
```suggestion
    return state->dev->PSR.reg & SDHC_PSR_CARDINS;
```

**Comment by dylad** [comment]
@benpicco @kfessel Third round is done.

**Review by kfessel — COMMENTED** [review]
looks fine testing?

**Comment by dylad** [comment]
I've run basics tests (timers, RTC, I2C, SPI, UART, USB, PWM) on `saml11-xpro`, `saml21-xpro`, `samr21-xpro`. I didn't spot any regression so far.

**Comment by dylad** [comment]
@benpicco @kfessel May I squash this one ?

**Comment by kfessel** [comment]
please squash

**Comment by dylad** [comment]
Squashed !

**Comment by dylad** [comment]
Fixed a typo in the UART driver spotted by Murdock.
Let's see if it's happy now.

**Comment by dylad** [comment]
All green !

**Review by kfessel — APPROVED** [review]
translation looks correct,

For the test i trust @dylad.

**Comment by dylad** [comment]
Murdock spotted an inconsistency in PWM driver for SAMD20. That should be good now.

**Comment by dylad** [comment]
Thanks !

**Pull request merged**